### ISSUE TITLE
local_useless_alloc yields incorrect substitution + hides shape errors

### ISSUE THREAD

**Issue by sisp** [Theano/Theano#2584]
I've noticed that `local_useless_alloc` yields incorrect results and hides shape errors. Consider the following example:
```
import theano
import theano.tensor as T

x = T.matrix('x')
y = T.alloc(x, 3, x.shape[1])
f = theano.function([x], y)
theano.printing.pydotprint(f, 'graph1.png')
```
This is the resulting graph:
![graph1](https://cloud.githubusercontent.com/assets/2206639/6573541/32ba646c-c71c-11e4-874b-974149f5944a.png)
This graph has two problems:
* It yields an incorrect result when `x` has shape (1, *) since the first dimension is not repeated.
* It hides shape errors when `x.shape[0] != 3`.

**Comment by lamblin** [comment]
Both problems are really the same: an `(1, n)` matrix should not be repeated, because its first dimension is not broadcastable, and it should be an error.
We try to catch those errors when we can, but as an optimizing compiler, Theano sometimes assumes that the original computation makes sense, and will hide errors. The way to catch them is then to compile the function without optimizations, or in `DebugMode`.
However, only the second case raises an error in `DebugMode`, which is worrying.

**Comment by sisp** [comment]
How about adding an `Assert` which checks the shapes of non-broadcastable dimensions?

**Comment by lamblin** [comment]
Yes, that would be an option.
One concern we used to have with asserts is that they could force the inclusion of large computations in the graph, just for the purpose of asserting the result of those computations.
However, we know have "unsafe" optimizations that remove asserts if everything is correct, so that would be possible in that case.

**Comment by sisp** [comment]
On second thought, however, `Assert` is a runtime check, so it would just throw a different exception at runtime.

Perhaps `Alloc` should throw an exception at runtime, when an array is passed whose non-broadcastable dimension does not match the corresponding shape value passed to `Alloc` at graph build time. How about that?

**Comment by lamblin** [comment]
Yes, an additional check in `Alloc.perform` would be a good idea as well for the first case. The second case is already failing when the `Alloc` is still there.
If we want to have a check even if the optimization has been applied, then we would need an Assert to be inserted by the optimization (which also makes sense).

**Comment by nouiz** [comment]
Do gh-2521 fix this? I think so, but can you confirm?

**Comment by sisp** [comment]
Do you mean https://github.com/Theano/Theano/pull/2586?

**Comment by nouiz** [comment]
I mean gh-2521. I think the problem was that we didn't included assert in the graph. But I could be wrong and this could be fixed by gh-2586. Do you know if any of those PR fix the problem in this issue?

**Comment by sisp** [comment]
I ran the above code using both PRs, but the graphs look the same as the one above. If, at all, I'd have expected https://github.com/Theano/Theano/pull/2586 could have made a difference because the other PR is only related to indexing. `local_useless_alloc` simply removes the `Alloc` op when the input and output types are the same which causes non-broadcastable dimensions with shape 1 to be ignored when they are intended to be repeated by the user. It depends on whether you want to allow non-broadcastable dimensions with shape 1 to be repeatable. So as @lamblin said, `Alloc.perform` could have an additional check and `local_useless_alloc` could insert an `Assert` op.
